polygeist/tools/cgeist/Lib/ValueCategory.cc

```diff
@@ -107,8 +107,12 @@ void ValueCategory::store(mlir::OpBuilder &builder, mlir::Value toStore) const {
           llvm::errs() << "lhs: " << val << "rhs: " << toStore << "\n";
           llvm::errs() << "lhs: " << LHS << "rhs: " << RHS << "\n";
         }
-        toStore =
-            builder.create<polygeist::Memref2PointerOp>(loc, LHS, toStore);
+        auto updatedPtrTy = LLVM::LLVMPointerType::get(
+            LHS.getElementType(), RHS.getMemorySpaceAsInt());
```

Comment:
RHS -> LHS   and pls add an assert that LHS and RHS have to have the same element type and address space.

Reply:
This is strange, I am thinking that any necessary bitcast (to change the element type) or addrspacecast (to change the address space) should be in the code *already*. Can you pls check whether some other operation upstream of this store should have added the address space cast first. Basically I think is good to dig deeper and then we can revisit this code once you have more info ?